src/traced/probes/packages_list/packages_list_data_source.cc

```diff
@@ -16,89 +16,201 @@
 
 #include "src/traced/probes/packages_list/packages_list_data_source.h"
 
+#include "perfetto/base/task_runner.h"
+#include "perfetto/base/time.h"
 #include "perfetto/ext/base/scoped_file.h"
 #include "perfetto/ext/base/string_splitter.h"
 
 #include "perfetto/ext/tracing/core/trace_writer.h"
 #include "protos/perfetto/trace/trace_packet.pbzero.h"
 
+#include "src/traced/probes/common/android_cpu_per_uid_poller.h"
 #include "src/traced/probes/packages_list/packages_list_parser.h"
 
 using perfetto::protos::pbzero::PackagesListConfig;
 
 namespace perfetto {
 
+namespace {
+constexpr uint32_t kFirstPackageUid = 10000;
+constexpr uint32_t kMinPollIntervalMs = 60000;
+constexpr uint32_t kWriteAllAtStart = 0xffffffff;
+}  // namespace
+
 // static
 const ProbesDataSource::Descriptor PackagesListDataSource::descriptor = {
     /*name*/ "android.packages_list",
     /*flags*/ Descriptor::kFlagsNone,
     /*fill_descriptor_func*/ nullptr,
 };
 
-bool ParsePackagesListStream(protos::pbzero::PackagesList* packages_list_packet,
+bool ParsePackagesListStream(base::FlatHashMap<uint64_t, Package>& packages,
```

Comment:
There are many packages with the same UID (I had a quick check on my acloud device), I don't think we can use a map like this here.